Content.Server/GameObjects/Components/Power/NewPower/BasePowerNetConnector.cs

```diff
@@ -0,0 +1,151 @@
+using System.Diagnostics;
+using System.Linq;
+
+namespace Content.Server.GameObjects.Components.NewPower
+{
+    /// <summary>
+    ///     Joins a <see cref="PowerNet"/> when created, allowing them to interact with the
+    ///     power system.
+    /// </summary>
+    public abstract class BasePowerNetConnector : Component
+    {
+        /// <summary>
+        ///     Connectors can only connect to the powernet of wires with the same voltage.
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)]
+        public Voltage Voltage { get => _voltage; set => SetVoltage(value); }
+        private Voltage _voltage;
+
+        /// <summary>
+        ///     The powernet this connector is currently a member of.
+        /// </summary>
+        [ViewVariables]
+        public PowerNet PowerNet { get; private set; }
+
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);
+            serializer.DataField(ref _voltage, "voltage", Voltage.High);
+        }
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            EnsureHasPowerNet();
+        }
+
+        public override void OnRemove()
+        {
+            LeavePowerNet();
+            base.OnRemove();
+        }
+
+        /// <summary>
+        ///     If this connector is not in a PowerNet, causes this connector to join a PowerNet
+        ///     on an adjacent wire or make one for itself.
+        /// </summary>
+        /// <returns>Returns whether or not this connector needed a powernet.</returns>
+        public bool EnsureHasPowerNet()
+        {
+            if (PowerNet != null)
+                return false;
+
+            //try to join an adjacent powernet
+            var nearbyWireNets = Owner.GetComponent<SnapGridComponent>()
+                .GetCardinalNeighborCells()
+                .SelectMany(sgc => sgc.GetLocal())
+                .Select(entity => entity.TryGetComponent<PowerWireComponent>(out var connector) ? connector : null)
+                .Where(connector => connector != null)
+                .Select(connector => connector.PowerNet)
+                .Where(wireNet => wireNet != null);
+
+            foreach (var wireNet in nearbyWireNets)
+            {
+                if (TrySetPowerNet(wireNet))
+                    break;
+            }
+            //otherwise make own powernet
+            if (PowerNet == null)
+            {
+                Debug.Assert(TrySetPowerNet(new PowerNet(this)));
+            }
+            return true;
+        }
+
+        /// <summary>
+        ///     Sets this connector's powernet reference if it sucessfully joins the supplied
+        ///     powernet.
+        /// </summary>
+        /// <returns>Returns whether or not this could join the supplied powernet.</returns>
+        public bool TrySetPowerNet(PowerNet powerNet)
+        {
+            if (TryJoinPowerNet(powerNet))
```

Comment:
This is a confusing naming, at least for me. I would expect `TryJoin` to cause the component to join the net. Calling `Leave` *after* that, but before actually joining, seems very unintuitive. Would it make sense to change `TryJoin` to `CanJoin`?